lib/at_cmd/at_cmd.c

```diff
@@ -162,133 +230,126 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 				"thread killed", errno);
 			close(common_socket_fd);
 			return;
-		} else if (item->data[bytes_read - 1] != '\0') {
+		} else if (buf[bytes_read - 1] != '\0') {
 
 			LOG_ERR("AT message too large for reception buffer or "
 				"missing termination character");
-
+			ret.state = AT_CMD_ERROR_READ;
 			ret.code  = -ENOBUFS;
 			goto next;
 		}
 
-		LOG_DBG("at_cmd_rx %d bytes, %s",
-			bytes_read, log_strdup(item->data));
+		LOG_DBG("at_cmd_rx %d bytes, %s", bytes_read, log_strdup(buf));
 
-		payload_len = get_return_code(item->data, &ret);
-
-		if (ret.state != AT_CMD_NOTIFICATION) {
-			if ((response_buf_len > 0) &&
-			    (response_buf != NULL)) {
-				if (response_buf_len >= payload_len) {
-					memcpy(response_buf, item->data,
-					       payload_len);
-				} else {
-					LOG_ERR("Response buffer not large "
-						"enough");
-
-					ret.code  = -EMSGSIZE;
-				}
-
-				response_buf_len = 0;
-				response_buf     = NULL;
+		payload_len = get_return_code(buf, &ret);
 
+		/* Verify the buffer size if provided, and copy the message */
+		if (current_cmd.cmd != NULL &&
+		    current_cmd.resp != NULL &&
+		    ret.state != AT_CMD_NOTIFICATION) {
+			if (current_cmd.resp_size < payload_len) {
+				LOG_ERR("Response buffer not large enough");
+				ret.code  = -EMSGSIZE;
 				goto next;
 			}
+			memcpy(current_cmd.resp, buf, payload_len);
 		}
 
-		if (payload_len == 0) {
-			goto next;
+		/* Call the relevant callback, if any */
+		if (ret.state == AT_CMD_NOTIFICATION &&
+		    notification_handler != NULL) {
+			notification_handler(buf);
```

Comment:
Are you sure you want to run this in the same thread and not put it on the work queue? This may end up in a deadlock of the notification handler is sending another AT command with at_cmd_write();

Reply:
It is at least preferred. It implements https://github.com/nrfconnect/sdk-nrf/pull/2264/files/c3e874cda311323d5ca7f825bab95f45ba063b6f#pullrequestreview-412504093 to save significant memory and avoid needing the heap.
The deadlock potential is documented and crashes out at at_cmd:319 just in case:
```c
	__ASSERT(k_current_get() != socket_tid,
		 "at_cmd deadlock: socket thread blocking self\n");
```